Fix missing setlocale with php 8

YoitoFes opened the pull request:
When php version = 8, basename('§') returns '§' even if LC_ALL is non-UTF-8 locale.
Since OC_Util::isSetLocaleWorking() assumes basename('§') returns empty string with non-UTF-8 locale,
the function  with php 8 incorrectly skips setlocale("C.UTF-8").

Fix it by using escapeshellcmd instead of basename.

This issue found in #29296 and The PR may be able to resolve it.

YoitoFes committed:
Fix missing setlocale with php 8

When php version = 8, basename('§') does not bug even if LC_ALL is non-UTF-8 locale.
This cause OC_Util::isSetLocaleWorking() to skip setlocale("C.UTF-8").

Fix it by using escapeshellcmd instead of basename.

Signed-off-by: Naoto Kobayashi <naoto.kobayashi4c@gmail.com>

YoitoFes committed:
Check whether setlocale works only after setlocale

Signed-off-by: Naoto Kobayashi <naoto.kobayashi4c@gmail.com>

PVince81 commented on lib/private/legacy/OC_Util.php:
```diff
@@ -1245,16 +1245,17 @@ public function isHtaccessWorking(\OCP\IConfig $config) {
 	 */
 	public static function isSetLocaleWorking() {
-		if ('' === basename('§')) {
+		if ('' === escapeshellcmd('§')) {
```
should we keep both checks, just in case ?
otherwise we'd need to check if `escapeshellcmd` behaves the same as `basename` in all past supported PHP versions

YoitoFes replied:
@PVince81 
Thank you for your feedback!

I added the `testIsSetLocaleWorking()` to PHPUnit. By using the test, GitHub Actions/Drone CI can check if `escapeshellcmd` behaves the same as `basename` in all supported PHP.

PVince81 replied:
indeed, that's clever

YoitoFes committed:
Add check whether escapeshellcmd behaves the same as basename

Signed-off-by: Naoto Kobayashi <naoto.kobayashi4c@gmail.com>

PVince81 reviewed (approved):
:+1:

PVince81 commented:
/backport to stable23

PVince81 commented:
/backport to stable22

PVince81 commented:
/backport to stable21

acsfer commented:
What if `escapeshellcmd()` is a disabled function in php? 🤔

PVince81 commented:
> What if `escapeshellcmd()` is a disabled function in php? thinking

not sure if it's as common as disabling the `exec` and `shell_exec`

but you're right to be careful, is there a way to detect disabled functions ?
if disabled then we could fall back to `basename`

acsfer commented:
@PVince81 https://www.php.net/manual/en/function.function-exists.php
```
if (!function_exists('escapeshellcmd')) {
...
}
```

YoitoFes commented:
Is `escapeshellarg` disabled less often than `escapeshellcmd`? `escapeshellarg` is used in Nextcloud's source code while `escapeshellcmd` is not.

acsfer commented:
Hard to say, but it is always a good practice check if the function exists before call it, and have a fallback 😉

YoitoFes committed:
OC_Util: Add fallbacks to check if current locale is UTF8

Using escapeshellcmd to get current locale causes error
if the function is disabled.

Add fallbacks to prevent the error.

Signed-off-by: Naoto Kobayashi <naoto.kobayashi4c@gmail.com>

YoitoFes committed:
OC_Util::isSetLocaleWorking: fix typo

Signed-off-by: Naoto Kobayashi <naoto.kobayashi4c@gmail.com>

YoitoFes commented:
I think `basename` can be used as a fallback as the function with php8 cause the issue.
So I added `escapeshellarg` and `preg_match` as fallbacks.

PVince81 reviewed (approved):
Fine by me :+1:

acsfer commented on lib/private/legacy/OC_Util.php:
```diff
@@ -1239,22 +1239,38 @@ public function isHtaccessWorking(\OCP\IConfig $config) {
+	private static function isNonUTF8Locale() {
+		if (function_exists("escapeshellcmd")) {
+			return ('' === escapeshellcmd('§'));
+		} else if (function_exists("escapeshellarg")) {
```
`elseif` 😝

YoitoFes committed:
OC_Util::isNonUTF8Locale: fix lint error

Signed-off-by: Naoto Kobayashi <naoto.kobayashi4c@gmail.com>

PVince81 commented:
seems the bot is lagging, no further backports yet?

PVince81 commented:
/backport to stable22

PVince81 commented:
/backport to stable21
